src/middlewared/middlewared/plugins/ldap.py

```diff
@@ -0,0 +1,639 @@
+    LEAVING = 2
+    JOINING = 3
+    HEALTHY = 4
+
+
+class SSL(enum.Enum):
+    NOSSL = 'off'
+    USESSL = 'on'
+    USETLS = 'start_tls'
+
+
+class LDAPQuery(object):
+    def __init__(self, **kwargs):
+        super(LDAPQuery, self).__init__()
+        self.ldap = kwargs.get('conf')
+        self.logger = kwargs.get('logger')
+        self.hosts = kwargs.get('hosts')
+        self.pagesize = 1024
+        self._isopen = False
+        self._handle = None
+        self._rootDSE = None
+        return
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        if self._isopen:
+            self._close()
+
+    def validate_credentials(self):
+        """
+        :validate_credentials: simple check to determine whether we can establish
+        an ldap session with the credentials that are in the configuration.
+        """
+        ret = self._open()
+        if ret:
+            self._close()
+        return ret
+
+    def _open(self):
+        """
+        We can only intialize a single host. In this case,
+        we iterate through a list of hosts until we get one that
+        works and then use that to set our LDAP handle.
+
+        SASL GSSAPI bind only succeeds when DNS reverse lookup zone
+        is correctly populated. Fall through to simple bind if this
+        fails.
+        """
+        res = None
+        if self._isopen:
+            return True
+
+        if self.hosts:
+            saved_simple_error = None
+            saved_gssapi_error = None
+            for server in self.hosts:
+                proto = 'ldaps' if SSL(self.ldap['ssl']) == SSL.USESSL else 'ldap'
+                port = 636 if SSL(self.ldap['ssl']) == SSL.USESSL else 389
+                uri = f"{proto}://{server}:{port}"
+                try:
+                    self._handle = ldap.initialize(uri)
+                except Exception as e:
+                    self.logger.debug(f'Failed to initialize ldap connection to [{uri}]: ({e}). Moving to next server.')
+                    continue
+
+                res = None
+                ldap.protocol_version = ldap.VERSION3
+                ldap.set_option(ldap.OPT_REFERRALS, 0)
+                ldap.set_option(ldap.OPT_NETWORK_TIMEOUT, 10.0)
+
+                if SSL(self.ldap['ssl']) != SSL.NOSSL:
+                    ldap.set_option(ldap.OPT_X_TLS_ALLOW, 1)
+                    ldap.set_option(
+                        ldap.OPT_X_TLS_CACERTFILE,
+                        f"/etc/certificates/{self.ldap['certificate']['cert_name']}.crt"
+                    )
+                    ldap.set_option(
+                        ldap.OPT_X_TLS_REQUIRE_CERT,
+                        ldap.OPT_X_TLS_ALLOW
+                    )
+
+                if SSL(self.ldap['ssl']) == SSL.USESSL:
+                    try:
+                        self._handle.start_tls_s()
+
+                    except ldap.LDAPError as e:
+                        self.logger.debug('%s', e)
```

Comment:
Should you `continue` if we hit this error?

A more descriptive debug error message would be useful as well